src/raftstore/store/worker/apply.rs

```diff
@@ -208,78 +254,114 @@ impl ApplyDelegate {
             }
 
             let res = match entry.get_entry_type() {
-                EntryType::EntryNormal => self.handle_raft_entry_normal(entry),
-                EntryType::EntryConfChange => self.handle_raft_entry_conf_change(entry),
+                EntryType::EntryNormal => self.handle_raft_entry_normal(apply_ctx, entry),
+                EntryType::EntryConfChange => self.handle_raft_entry_conf_change(apply_ctx, entry),
             };
 
             if let Some(res) = res {
                 results.push(res);
             }
         }
 
+        if !self.pending_remove {
+            self.write_apply_state(apply_ctx.wb_mut());
+        }
+
+        self.metrics.written_bytes += apply_ctx.wb_ref().data_size() as u64;
+        self.metrics.written_keys += apply_ctx.wb_ref().count() as u64;
+
         slow_log!(t,
                   "{} handle ready {} committed entries",
                   self.tag,
                   committed_count);
         results
     }
 
-    fn handle_raft_entry_normal(&mut self, entry: Entry) -> Option<ExecResult> {
+    fn write_apply_state(&self, wb: &mut WriteBatch) {
```

Comment:
This may be reasonable or unreasonable, many other libraries also use this way for the FFI API, you can even see some Rust API uses immutable self too, like TcpStream `SetXXX`. So I think this is ok here, and we should follow the style and keep consistent now.